design/src/rfcs/rfc0022_error_context_and_compatibility.md

```diff
@@ -0,0 +1,370 @@
+}
+```
+
+Each error variant gets its own struct, which can hold error-specific contextual information.
+Both the error enum and the details on each variant are extensible.
+
+The code generated errors are already aligned with the goals for this RFC.
+
+Approaches from other projects
+------------------------------
+
+### `std::io::Error`
+
+The standard library uses an `Error` struct with an accompanying `ErrorKind` enum
+for its IO error. Roughly:
+
+```rust
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum ErrorKind {
+    NotFound,
+    // ... omitted ...
+    Other,
+}
+
+#[derive(Debug)]
+pub struct Error {
+    kind: ErrorKind,
+    source: Box<dyn std::error::Error + Send + Sync>,
+}
+```
+
+What this error does well:
+- It is extensible since the `ErrorKind` is non-exhaustive
+- It has an `Other` error type that can be instantiated by users in unit tests,
+  making it easier to unit test error handling
+
+What could be improved:
+- There isn't an ergonomic way to add programmatically accessible error-specific context
+  to this error in the future
+- The source error can be downcasted, which could be a trap for backwards compatibility.
+
+### Hyper 1.0
+
+Hyper is has outlined [some problems they want to address with errors](https://github.com/hyperium/hyper/blob/bd7928f3dd6a8461f0f0fdf7ee0fd95c2f156f88/docs/ROADMAP.md#errors)
+for the coming 1.0 release. To summarize:
+
+- It's difficult to match on specific errors (Hyper 0.x's `Error` relies
+  on `is_x` methods for error matching rather than enum matching).
+- Error reporters duplicate information since the hyper 0.x errors include the display of their error sources
+- `Error::source()` can leak internal dependencies
+
+Opaque Error Sources
+--------------------
+
+There is [discussion in the errors working group](https://github.com/rust-lang/project-error-handling/issues/53)
+about how to avoid leaking internal dependency error types through error source downcasting. One option is to
+create an opaque error wrapping new-type that removes the ability to downcast to the other library's error.
+This, however, can be circumvented via unsafe code, and also breaks the ability for error reporters to
+properly display the error (for example, if the error has backtrace information, that would be
+inaccessible to the reporter).
+
+This situation might improve if the nightly `request_value`/`request_ref`/`provide` functions on
+`std::error::Error` are stabilized, since then contextual information needed for including things
+such as a backtrace could still be retrieved through the opaque error new-type.
+
+This RFC proposes that `Box<dyn Error + Send + Sync + 'static>` is used to refer to errors from libraries.
+An allocation to store this can be avoided if the error source is private, in which case, a reference to it
+can be returned from the `Error::source` implementation. Otherwise, if it's directly exposed as a public enum
+field, then it must be boxed.
```

Comment:
I'm not confident that I got this right. Perhaps we should take a harder stance and make a rule to not directly expose the source error (outside of the `Error::source` call) so that we can replace it with an opaque error type later when that becomes more practical. The downside is that every enum variant that has a source error will need to have a separate struct in the same namespace in order to hide it.

Would love to get more input on this.

Reply:
To summarize an offline discussion with @Velfi: We agreed the error source should only be available through `Error::source`, and not via a public enum field. This means that each enum struct needs to be an actual struct so that it can have private members. To avoid obscuring the docs with a bunch of error structs, each module that has errors will need to have those errors moved into a new `error` submodule so that the error structs are grouped together and can't be confused with other non-error structs in the documentation.